<launch>
  <!-- Stopline Detection Node -->
  <node name="stopline_detection_node" pkg="stopline_pkg" type="stopline_detection_node.py">
    <!-- 이미지 크롭 파라미터 -->
    <param name="crop_top" value="280" />

    <!-- 정지선 감지 파라미터 -->
    <param name="down_hist_start_line" value="0" />
    <param name="stopline_threshold" value="5" />
    <rosparam param="count_thresholds">[3, 5, 5, 7, 3, 5, 1, 4, 5, 5]</rosparam>
    <rosparam param="cooldown_durations">[3.0, 10.0, 5.0, 15.0, 1.5, 3.0]</rosparam>
    <param name="show_debug_image" value="true" />

    <!-- 히스토그램 임계값 파라미터 -->
    <param name="histogram_threshold" value="200" />

    <!-- Otsu 임계값 조정 파라미터 -->
    <param name="otsu_threshold_offset" value="25" />

    <!-- 허프라인 파라미터 -->
    <param name="hough_threshold" value="50" />
    <param name="hough_min_line_length" value="40" />
    <param name="hough_max_line_gap" value="10" />
    <param name="hough_angle_tolerance" value="25.0" />
    <param name="hough_use_morphology" value="true" />

    <!-- 색공간 선택 파라미터 -->
    <param name="use_lab_colorspace" value="false" />

    <!-- 노란색 정지선 감지 파라미터 (state 9용) -->
    <rosparam param="yellow_hsv_lower">[0, 40, 50]</rosparam>
    <rosparam param="yellow_hsv_upper">[26, 110, 255]</rosparam>
    <param name="yellow_state" value="9" />
    <param name="log_level" value="debug"/>
  </node>

    <!-- 차선 검출 및 조향 제어 노드 -->
  <node name="onelane_detection" pkg="onelane_detection" type="onelane_detection.py">
      <!-- 카메라 이미지 토픽 -->
      <param name="image_topic" value="/usb_cam/image_rect_color" />
      
      <!-- 제어 명령 발행 토픽 -->
      <param name="cmd_topic" value="/camera/steering" />
`
      <!-- 차선 선택: "left" 또는 "right" -->
      <param name="lane_side" value="right" />
      
      <!-- Sliding Window 알고리즘 파라미터 -->
      <param name="window_num" value="12" />
      <param name="window_margin" value="30" />
      <param name="crop_start_ratio" value="0.5" />
      
      <!-- HSV 색상 필터링 임계값 (노란색 차선 기준) -->
      <rosparam param="hsv_lower">[3, 100, 35]</rosparam>
      <rosparam param="hsv_upper">[40, 255, 255]</rosparam>
      
      <!-- 적응형 이진화 파라미터 -->
      <param name="adaptive_block_size" value="15" />
      <param name="adaptive_c" value="-1" />
      
      <!-- 조향 기준점 비율 (0.0~1.0) -->
      <!-- <param name="steering_ratio" value="0.8" /> -->
      <!-- 왼쪽/오른쪽 차선별 개별 설정 (옵션) -->
      <param name="steering_ratio_left" value="0.3" />
      <param name="steering_ratio_right" value="0.7" />
      
      <!-- 조향 게인 -->
      <!-- 통합 게인 (설정 시 좌우 구분 무시) -->
      <!-- <param name="steering_gain" value="0.66" /> -->
      <!-- 왼쪽/오른쪽 차선별 개별 게인 -->
      <param name="steering_gain_left" value="0.66" />
      <param name="steering_gain_right" value="0.55" />
      
      <!-- 주행 제어 -->
      <param name="enable_drive" value="true" />
      
      <!-- 디버그 창 표시 여부 -->
      <param name="show_debug" value="true" />
      
      <!-- mission_state별 차선 방향 매핑 (옵션) -->
      <!-- 예: state 0,2,4는 오른쪽, state 1,3,5는 왼쪽 차선 감지 -->
      <rosparam param="lane_direction_map">{"3": "left", "6": "right"}</rosparam>
  </node>
</launch>
